### PR TITLE
Basic support for UD

### PR DESCRIPTION
**Pull request by martinpopel**
Proof of concept (for benchmarks) of a support for Universal Dependencies and CoNLL-U format

### CODE REVIEW

File: pytreex/block/read/conllu.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        zone = bundle.create_zone(self.language, self.selector)
+        root = zone.create_atree()
+        last_node = root
+        nodes = [root]
+        parents = [0]
+        
+        for line in fh:
+            
+            # Strip newline character
+            line = line.rstrip('\n')
```
Comment (tuetschek) [inline pytreex/block/read/conllu.py]: How about \r?

File: pytreex/block/read/conllu.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                root = zone.create_atree()
+                last_node = root
+                nodes = [root]
+                parents = [0]
+                continue
+            
+            columns = line.split('\t')
+            
+            # TODO: multi-word tokens
+            if columns[0].find('-') >= 0: continue
```
Comment (tuetschek) [inline pytreex/block/read/conllu.py]: Better written `if '-' in columns[0]`

**Comment by martinpopel** [comment]
> HEAD field being an underscore

I use `parents.append(int(columns[6]))`, so underscore as HEAD results in
ValueError: invalid literal for int() with base 10: '_'
This is not very helpful error message for the user, but at least it does show that there is some error in the conllu file.
We will need anyway some kind of validator (different from the reader) for CoNLL-U, which will indicate the line (and context) where the error was and which will detect also some minor errors (e.g. FEATS not sorted alphabetically) that are not fatal for the reader.

**Comment by foxik** [comment]
> I use parents.append(int(columns[6])), so underscore as HEAD results in
> `ValueError: invalid literal for int() with base 10: '_'`
> This is not very helpful error message for the user, but at least it does show that there is some error in the conllu file.

TL;DR I think we should graciously handle underscore in HEAD.

Even if according to CoNLL-U format documentation head should be node ID or 0, I think underscore in the HEAD column is a perfectly reasonable value. Consider for example a pipeline, which starts by tokenizing the text and dumping the CoNLL-U. Underscore will be used in many other positions, so it may appear in HEAD too, when you want to say that "no HEAD value is set" (which is different than 0, which means an edge to root node is present). Also multi-word tokens explicitly allow underscores in HEAD, which will make people using them for words too.

**Comment by tuetschek** [comment]
I agree with @foxik . @martinpopel , are you afraid of reduced performance? Or is there another reason to not allow underscores?

**Comment by foxik** [comment]
One of the issues is how to represent underscore on head position. Can a node _not_ contain `.parent` attribute? Or maybe have `.parent` set to `None`?

**Comment by tuetschek** [comment]
Oh, sorry, I forgot about this... I thought about hanging under root in that case.

If a node has `.parent` set to `None`, it's supposed to be a root. And a single bundle can only have one root of one kind, so there's now no way of telling that "node X belongs to tree Y, but we don't know where". And I don't think this would work in Perl Treex either.

I think that defaulting to hanging under root would be a relatively clean solution...

**Comment by foxik** [comment]
Nevertheless in that case the head will be printed as 0?

I think that if we accept underscores as heads, we should also print them as underscores, which will not be straightforward with current representation. We can either add a boolean value indicating this (and resetting it when we change parent of the node), or maybe disallow underscores alltogether if we are unable to keep them intact.

**Comment by martinpopel** [comment]
In old Treex, Pytreex and several other implementations from https://github.com/martinpopel/newtreex, there is no way of distinguishing a node with unknown parent and a node under root. The philosophy of UD (at least from the Prague/PML point of view) is to keep things simple. So I think there is no need of distinguishing this. You can use CoNLL-U for tagging and keep the HEAD as 0 (except for multiwords). I agree we can make the reader more robust, so it interprets _ as 0 at the HEAD field.

**Comment by foxik** [comment]
If you both agree, then why not, let _ in head mean 0.

(In my tools, I will keep it as _, but that is undistinguishable for your tools :-)

**Comment by martinpopel** [comment]
> let _ in head mean 0.

OK, it's there, the reader is more robust now. I still think that '_' (or any other non-integer) as HEAD makes the CoNLL-U file invalid. Other tools may not support such files, so we should not promote it. Maybe be we can read such files, but produce a warning (only for the first occurrence in each file). But as I wrote, we should have a separate validator anyway.
